src/LinearSolvers/ConjugateGradientSolver.jl

```diff
@@ -0,0 +1,469 @@
+    # tolerances (2)
+    rtol::FT
+    atol::FT
+    # arrays of size reshape_tuple (7)
+    r0::AT1
+    z0::AT1
+    p0::AT1
+    r1::AT1
+    z1::AT1
+    p1::AT1
+    Lp::AT1
+    # arrays of size(MPIStateArray) which are aliased to two of the previous dimensions (2)
+    alias_p0::AT2
+    alias_Lp::AT2
+    # reduction dimension (1)
+    dims::RD
+    # reshape dimension (1)
+    reshape_tuple::RT
+    # maximum number of iterations (1)
+    max_iter::IT
+
+end
+
+# Define the outer constructor for the ConjugateGradient struct
+"""
+function ConjugateGradient(Q::AT; rtol = eps(eltype(Q)), atol = eps(eltype(Q)), dims = :) where AT
+
+# Description
+- Outer constructor for the ConjugateGradient struct
+
+# Arguments
+- `Q`:(array). The kind of object that linearoperator! acts on.
+
+# Keyword Arguments
+- `rtol`: (float). relative tolerance
+- `atol`: (float). absolute tolerance
+- `dims`: (tuple or : ). the dimensions to reduce over
+- `reshape_tuple`: (tuple). the dimensions that the conjugate gradient solver operators over
+
+# Comment
+- The reshape tuple is necessary in case the linearoperator! is defined over vectors of a different size as compared to what plays nicely with the dimension reduction in the ConjugateGradient. It also allows the user to define preconditioners over arrays that are more convenienently shaped.
+
+# Return
+- ConjugateGradient struct
+"""
+function ConjugateGradient(
+    Q::AT;
+    rtol = eps(eltype(Q)),
+    atol = eps(eltype(Q)),
+    max_iter = length(Q),
+    dims = :,
+    reshape_tuple = size(Q),
+) where {AT}
+
+    # create empty container for pushing struct objects
+    container = []
+    # allocate arrays (5)
+    r0 = reshape(similar(Q), reshape_tuple)
+    z0 = reshape(similar(Q), reshape_tuple)
+    r1 = reshape(similar(Q), reshape_tuple)
+    z1 = reshape(similar(Q), reshape_tuple)
+    p1 = reshape(similar(Q), reshape_tuple)
+    # allocate array of different shape (2)
+    alias_p0 = similar(Q)
+    alias_Lp = similar(Q)
+    # allocate create aliased arrays (2)
+    p0 = reshape(alias_p0, reshape_tuple)
+    Lp = reshape(alias_Lp, reshape_tuple)
+    # push to container (2)
+    push!(container, rtol)
+    push!(container, atol)
+    # push to container (7)
+    push!(container, r0)
+    push!(container, z0)
+    push!(container, p0)
+    push!(container, r1)
+    push!(container, z1)
+    push!(container, p1)
+    push!(container, Lp)
+    # push to container (2)
+    push!(container, alias_p0)
+    push!(container, alias_Lp)
+    # push to container (1)
+    push!(container, dims)
+    # push to container (1)
+    push!(container, reshape_tuple)
+    # push to container (1)
+    push!(container, max_iter)
+    # create struct instance by splatting the container into the default constructor
+    return ConjugateGradient{
+        typeof(z0),
+        typeof(Q),
+        eltype(Q),
+        typeof(dims),
+        typeof(reshape_tuple),
+        typeof(max_iter),
+    }(container...)
+end
```

Comment:
Why is the tuple `container` created?  What is the advantage of creating it as apposed to directly passing the arguments into `ConjugateGradient` similar to what we do in the rest of the code?

Reply:
The tuple container is created just to have cleaner looking code. I could directly pass all the arguments directly into the default constructer for conjugate gradient, but typically when I do that I mess something up. I'm happy to conform to whatever we do in the rest of the CliMa code!

Reply:
I’d suggest switching from `push!` to using
```
args = [
...
]
```
to avoid dynamically changing the array size. I know it’s not a big array but the size is known, so this syntax works fine too and doesn’t have the dynamic memory allocations.